gap/semigroups/semicons.gi

```diff
@@ -785,6 +785,58 @@ function(filter, G, n)
   return ReesZeroMatrixSemigroup(G, mat);
 end);
 
+InstallMethod(SemilatticeFromDigraph,
+"for a digraph", [IsDigraph],
+function(digraph)
+  local red, top, im, max, n, S, x, i, j;
+
+  if not IsJoinSemilatticeDigraph(digraph) then
+    if IsMeetSemilatticeDigraph(digraph) then
+      return SemilatticeFromDigraph(DigraphReverse(digraph));
+    else
+      ErrorNoReturn("Semigroups: SemilatticeFromDigraph ",
+                    "usage,\n", "the argument must be a join semilattice ",
+                    "digraph or a meet semilattice digraph,");
+    fi;
+  fi;
+
+  red := DigraphReflexiveTransitiveReduction(digraph);
+  top := DigraphTopologicalSort(digraph);
+  # im[i] will store the image of the idempotent partial perm corresponding to
+  # vertex i of the arugment <digraph>
+  im         := [];
+  im[top[1]] := [];
+  max        := 1;
+
+  n := DigraphNrVertices(digraph);
+  # For each vertex, the corresponding idempotent has an image
+  # containing all images of idempotents below it.
+  for i in [2 .. n] do
+    im[top[i]] := Union(List(OutNeighboursOfVertex(red, top[i]), j -> im[j]));
+    # When there is only one neighbour, we must add a point to the image to
+    # distinguish the two idempotent partial perms.
+    if Length(OutNeighboursOfVertex(red, top[i])) = 1 then
+      Add(im[top[i]], max);
+      max := max + 1;
+    fi;
+  od;
+
+  # Add the generators one by one
+  S := Semigroup(PartialPerm(im[top[n]], im[top[n]]));
```

Comment:
You should be able to exactly work out a minimal generating set a priori. Specifically, the generators should be the sources of the `DigraphReflexiveTransitiveReduction`, right? Something like that.

Reply:
A line digraph will have only one source but every element is in a minimal generating set. E.g.
`Digraph([[1], [1, 2], [1, 2, 3]])`. I think you are right that it is possible to work out a generating set without using `ClosureSemigroup` to do the checking though.

Reply:
Good thinking! Yeah the correct property is something like 'join-irreducibility' or 'meet-irreducibility' or something. If you can't be bothered working it out then no worries, it can always be improved later, maybe leave a comment to say so.